Renderable builder locks

p12tic opened the pull request:
Hi,

Currently step locks can be renderables whereas builder locks can't. This pull request implements renderable support for builder locks.

Could you please look and decide whether this PR is good in principle? I'd like not to spend time refining the PR if the implementation goes to completely wrong direction :-)

Currently there are no tests for the new functionality, some existing tests have been broken by changes of internal APIs and some other things are missing. These all will be fixed on the next iteration if you like the PR.

Thanks a lot for your time!

tardyp commented:
I think this might not be a very good idea to start breaking the APIs like this.

I would like to have more comments on why you are doing each change. Maybe just inside github comments. 

can't we just take the lock at a time where it is easier to render them?

p12tic commented on master/buildbot/config.py:
```diff
@@ -900,7 +901,7 @@ def check_lock(lock):
 
         for b in self.builders:
-            if b.locks:
+            if b.locks and not IRenderable.providedBy(b.locks):
```
This check is needed because renderable locks are not known until builder start and we can't do any checks at config time for these locks.

p12tic commented on master/buildbot/process/build.py:
```diff
@@ -190,25 +194,23 @@ def getWorkerName(self):
         return self.workerforbuilder.worker.workername
     deprecatedWorkerClassMethod(locals(), getWorkerName)
 
-    def setupProperties(self):
```
This was moved to a static method so that we don't need an instance of a Build to know its properties. The BuildRequestDistributor needs the final build properties in order to figure out whether a certain build request will satisfy the lock constraits before creating a build for it.

p12tic commented on master/buildbot/process/build.py:
```diff
@@ -190,25 +194,23 @@ def getWorkerName(self):
 
         # from the SourceStamps, which have properties via Change
-        for change in self.allChanges():
+        sources = requests[0].mergeSourceStampsWith(requests[1:])
```
This is the way we compute Build sources in its __init__. If there are more places this is used we could move the sources setup to a separate function, but for now I think the amount of duplication of logic is too small.

p12tic commented on master/buildbot/process/build.py:
```diff
@@ -225,19 +227,22 @@ def setupOwnProperties(self):
             props.setProperty("codebase", source.codebase, "Build")
             props.setProperty("project", source.project, "Build")
 
-    def setupWorkerForBuilder(self, workerforbuilder):
```
Previously setupWorkerForBuilder did 2 things:
 - Figured out the worker properties to add to build properties
 - Assigned the worker to the build

Since the BuildRequestDistributor also needs worker properties to figure out whether a build can be started for specific build request, this method was split into two so that it's possible to retrieve worker properties without creating a build.

We could probably merge this method into setupProperties, because there's no reason for these methods to be separate.

p12tic commented:
Sorry for a late reply.

I've added several comments to the diff. When implementing this change I tried to reduce the amount of changes to the absolute minimum possible. The primary reason why we change the Build API is that before this PR there was no way to retrieve properties for a build without creating an instance of a Build. This is needed by the BuildRequestDistributor in order to figure out whether a certain build request will satisfy the lock constraits.

If API breakage is the primary concern I could rename the new static Build APIs to something else and bring back the non-static APIs as wrappers.

Thanks a lot!

p12tic commented on master/buildbot/process/builder.py:
```diff
@@ -261,10 +263,21 @@ def getAvailableWorkers(self):
 
-    def canStartWithWorkerForBuilder(self, workerforbuilder):
+    @defer.inlineCallbacks
+    def canStartWithWorkerForBuilder(self, workerforbuilder, buildrequests):
```
If we want absolutely preserve this API as is, buildrequests could be set to None by default and the method would fail if any of the locks were renderable when the method was called without build request information.

tardyp commented:
This looks pretty reasonable to me.

Now there is another exercise needed is the inevitable update of tests, this might show how much deep those APIs are used

p12tic commented:
I've made the following changes:
 - Split setup of the 'builddir' worker property into a separate method, as it's only available once a worker is attached, which is quite late in the case of latent workers.
 - Because of the above, the number of property setup methods in the Build class started to become confusing. So I've merged Build.setupProperties and Build.setupWorkerProperties into Build.setupPropertiesKnownBeforeBuildStarts. Both of these methods were used in two places throughout the codebase next to each other, so no reason for them to be separate.
 - There were comparatively large use of Builder.canStartWithWorkerForBuilder in tests and it looked like something external code might actually use, so I fixed compatibility with the old API. Now it shows a warning when the `buildrequests` parameter is not provided and in case a renderable lock is configured, an exception is thrown.
 - Added new tests for the new behaviors.
 - Documented and added release notes.

It looks like there are some minor issues remaining, I'll fix them soon. Aside from that, what other problems could you identify in the PR?

Thanks!

p12tic commented:
Ping :) The tests results are all green, for some reason github has missed an update.

tardyp commented:
Looks like good work! Please come up with more of those!
